client/lib/wp/sync-handler/test/index.js

```diff
@@ -0,0 +1,164 @@
+	postResponseWithNoHandle: {},
+	postResponseWithHandle: {
+		meta: {
+			page_handle: 'test'
+		}
+	},
+	postResponseNewHandle: {
+		meta: {
+			page_handle: 'test2'
+		}
+	}
+}
+
+describe( 'sync-handler', () => {
+	before( () => {
+		mockery.enable( {
+			warnOnReplace: false,
+			warnOnUnregistered: false
+		} );
+		mockery.registerMock( 'lib/localforage', localforageMock );
+		const handlerMock = ( params, callback ) => {
+			const key = generateKey( params );
+			callback( null, responseData[ key ] );
+			return responseData[ key ];
+		};
+
+		( { SyncHandler, generateKey, hasPaginationChanged } = require( '../' ) );
+		wpcom = new SyncHandler( handlerMock );
+	} );
+
+	beforeEach( () => {
+		responseData = {};
+		localData = {};
+	} );
+
+	after( function() {
+		mockery.disable();
+	} );
+
+	it( 'should call callback with local response', () => {
+		const { postRequest } = testRequestData;
+		const key = generateKey( postRequest );
+		const callback = sinon.spy();
+		localData[ key ] = { body: 'test' };
+		wpcom( postRequest, callback );
+		expect( callback.calledWith( null, 'test' ) );
+	} );
+
+	it( 'should call callback with request response', () => {
+		const { postRequest } = testRequestData;
+		const key = generateKey( postRequest );
+		const callback = sinon.spy();
+		responseData[ key ] = { body: 'test' };
+		wpcom( postRequest, callback );
+		expect( callback ).to.have.been.calledOnce;
+		expect( callback.calledWith( null, 'test' ) );
+	} );
+
+	it( 'should call callback twice with local and request responses', () => {
+		const { postRequest } = testRequestData;
+		const key = generateKey( postRequest );
+		const callback = sinon.spy();
+		localData[ key ] = { body: 'test1' };
+		responseData[ key ] = ( null, { body: 'test2' } );
+		wpcom( postRequest, callback );
+		expect( callback ).to.have.been.calledTwice;
+		expect( callback.calledWith( null, 'test1' ) );
+		expect( callback.calledWith( null, 'test2' ) );
+	} );
+
+	describe( 'generateKey', () => {
+		it( 'should return the same key for identical request', () => {
+			const { simpleRequest } = testRequestData;
+			const secondRequest = Object.assign( {}, simpleRequest );
+			const key1 = generateKey( simpleRequest );
+			const key2 = generateKey( secondRequest );
+			expect( typeof key1 ).to.equal( 'string' );
+			expect( key1 ).to.equal( key2 );
+		} );
+		it( 'should return unique keys for different requests', () => {
+			const { simpleRequest } = testRequestData;
+			const secondRequest = Object.assign( { query: '?filter=test' }, simpleRequest );
+			const key1 = generateKey( simpleRequest );
+			const key2 = generateKey( secondRequest );
+			expect( typeof key1 ).to.equal( 'string' );
+			expect( key1 ).to.not.equal( key2 );
+		} );
+	} );
+
+	describe( 'hasPaginationChanged', () => {
+		it( 'should not call hasPaginationChanged for non-whitelisted requests', () => {
+			const { simpleRequest } = testRequestData;
+			sinon.spy( hasPaginationChanged );
```

Comment:
We should move this to a `before` or `beforeEach` block, and then clean up in a respective `after` or `afterEach` block with:
```javascript
hasPaginationChanged.restore();
```